ide/src/main/java/com/devonfw/tools/ide/cli/Ide.java

```diff
@@ -0,0 +1,192 @@
+
+  /**
+   * The actualy main method of the CLI program.
+   *
+   * @param args the command-line arguments.
+   */
+  public static void main(String... args) {
+
+    int exitStatus = new Ide().run(args);
+    System.exit(exitStatus);
+  }
+
+  private IdeContext context() {
+
+    if (this.context == null) {
+      // fallback in case of exception before initialization
+      this.context = new IdeContextConsole(IdeLogLevel.INFO, null, false);
+    }
+    return this.context;
+  }
+
+  /**
+   * Non-static variant of {@link #main(String...) main method} without invoking {@link System#exit(int)} so it can be
+   * tested.
+   *
+   * @param args the command-line arguments.
+   * @return the exit code.
+   */
+  public int run(String... args) {
+
+    int exitStatus;
+    try {
+      exitStatus = runOrThrow(args);
+    } catch (CliException error) {
+      exitStatus = error.getExitCode();
+      if (context().level(IdeLogLevel.DEBUG).isEnabled()) {
+        context().error(error.getMessage(), error);
+      } else {
+        context().error(error.getMessage());
+      }
+    } catch (Throwable error) {
+      exitStatus = 255;
+      String title = error.getMessage();
+      if (title == null) {
+        title = error.getClass().getName();
+      } else {
+        title = error.getClass().getSimpleName() + ": " + title;
+      }
+      String message = "An unexpected error occurred!\n" //
+          + "We are sorry for the inconvenience.\n" //
+          + "Please check the error below, resolve it and try again.\n" //
+          + "If the error is not on your end (network connectivity, lack of permissions, etc.) please file a bug:\n" //
+          + "https://github.com/devonfw/ide/issues/new?assignees=&labels=bug&projects=&template=bug.md&title="
+          + URLEncoder.encode(title, StandardCharsets.UTF_8);
+      context().error(error, message);
+    }
+    return exitStatus;
+  }
+
+  /**
+   * Like {@link #run(String...)} but does not catch {@link Throwable}s so you can handle them yourself.
+   *
+   * @param args the command-line arguments.
+   * @return the exit code.
+   */
+  public int runOrThrow(String... args) {
+
+    CliArgument first = CliArgument.of(args);
+    CliArgument current = initContext(first);
+    for (Commandlet commandlet : this.context.getCommandletManager().getCommandlets()) {
+      boolean matches = apply(current, commandlet);
+      if (matches) {
+        this.context.debug("Running commandlet {}", commandlet);
+        if (commandlet.isIdeHomeRequired() && (this.context.getIdeHome() == null)) {
+          throw new CliException(this.context.getMessageIdeHome());
+        }
+        commandlet.run();
+        return ProcessResult.SUCCESS;
+      } else {
+        this.context.trace("Commandlet did not match");
+      }
+    }
+    // TODO print help properly
+    context().info("Usage: ide «args»");
+    return 1;
+  }
+
+  private CliArgument initContext(CliArgument first) {
```

Comment:
In general we seem not to be on the same page about writing JavaDoc for private things.
The method `initContext` will initialize the context. Okay, why we need to pass the `CliArgument` and get `CliArgument` back is not obvious from the signature. I can write JavaDoc for this but my PoV is that you can read the code if you want to know this. A problem with adding JavaDoc is that it might not get maintained if the code gets changed and internal private methods tend to change often. A wrong JavaDoc is worse than no JavaDoc.
Did you ever read the book about "clean code"? If no, I would strongly recommend it...
Anyhow in this case, I am happy to add the JavaDoc and address your feedback as actually the `@return` would be helpful.

Reply:
I'm asking for javadocs because of IDE autocomplete and documentation functionalities. The more (maintained) javadocs we have for methods like these, the more the IDE's will help us to get through faster. I don't like to read each method body to understand what my params need to be and what will be returned and how.
I've not read "clean code" yet. Might give it a go later.